### DISCUSSION TITLE
Captcha API time outs

### DISCUSSION THREAD

**Discussion by 737simpilot** [CIDRAM/CIDRAM#299] [Q&A] [answered]
Question: Is there a built in time out for captcha API retrieval? And if so, what does it fallback to if anything? Perhaps if two captchas were configured it'll fall back to one or the other?

**Comment by Maikuolan** [comment, marked as answer]
When CIDRAM communicates with the upstream server (whether for ReCaptcha or HCaptcha, all the same), the same instance of the "Request" class as is used for pretty much all other outbound requests is used to do so, and the default timeout the instance uses can be configured via the `default_timeout` directive (under `general` configuration), defaulting to 12 seconds.

As for communication between the browser and the upstream server, I have no idea about that. I haven't coded anything into CIDRAM in that regard, so I guess, whatever the browser itself defaults to, or whatever is set by the respective APIs, if anything has been set.

**Comment by 737simpilot** [comment]
That's kinda what I thought about the CIDRAM timeout. Wasn't sure if that had a bearing on the captcha API or not. 

Thanks again.

**Comment by 737simpilot** [comment]
I forgot to ask for clarification. If I were to have both captchas enabled and one times out, does the other get used?

**Reply by Maikuolan** [reply]
> I forgot to ask for clarification. If I were to have both captchas enabled and one times out, does the other get used?

Nope. If we're talking regarding communication between the CAPTCHA server and the end-user (i.e., between the server and the browser), CIDRAM wouldn't have any way to know whether the request had timed out to begin with, so no way to be able to make the decision about switching to a different API anyway. If we're talking regarding communication between CIDRAM and the CAPTCHA server (e.g., input received from user and attempting to verify that input against the CAPTCHA server), by that point in time, a CAPTCHA would've already been served, so switching over to a different API at that point wouldn't make any sense anyway, other than by having the browser perform an entirely fresh request and starting the whole request process right from the beginning.

**Comment by 737simpilot** [comment]
Sorry, Caleb. I should have been more elaborate on my inquiry. Yes, I'm strictly talking about CIDRAM and the CAPTCHA server API. My thinking is this: user loads the page and gets a CAPTCHA for what ever reason. But some unforeseen event out there in the land of the fragile Internet and ASN takeovers and what ever else causes the CAPTCHA API request server side to timeout. At this point, would the other CAPTCHA that's enabled in CIDRAM fire off? 

I suppose the answer would again be a fat no given your explanation.  In which case I'm thinking a fallback mechanism _might_ be prudent to have for API CAPTCHA timeouts server side. It doesn't have to be something as remote as ASN shenanigans or what ever. A timeout could fall from within one's server or host's infrastructure its self. 

Just thinking outside the box, and me being me, I don't like single points of failure in a lot of things I do. Computers, data, realm of survivalism and everything you can think of.
